internal/sidecar/service/encryptionkeyrotation.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2024 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"context"
+
+	kube "github.com/csi-addons/kubernetes-csi-addons/internal/kubernetes"
+	"github.com/csi-addons/kubernetes-csi-addons/internal/proto"
+	csiEncKeyRotation "github.com/csi-addons/spec/lib/go/encryptionkeyrotation"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+type EncryptionKeyRotationServer struct {
+	proto.UnimplementedEncryptionKeyRotationServer
+	ekrClient  csiEncKeyRotation.EncryptionKeyRotationControllerClient
+	kubeClient *kubernetes.Clientset
+}
+
+func NewEncryptionKeyRotationServer(
+	c *grpc.ClientConn, kc *kubernetes.Clientset,
+) *EncryptionKeyRotationServer {
+	return &EncryptionKeyRotationServer{
+		ekrClient:  csiEncKeyRotation.NewEncryptionKeyRotationControllerClient(c),
+		kubeClient: kc,
+	}
+}
+
+func (ekrs *EncryptionKeyRotationServer) RegisterService(server grpc.ServiceRegistrar) {
+	proto.RegisterEncryptionKeyRotationServer(server, ekrs)
+}
+
+func (ekrs *EncryptionKeyRotationServer) EncryptionKeyRotate(
+	ctx context.Context, req *proto.EncryptionKeyRotateRequest,
+) (*proto.EncryptionKeyRotateResponse, error) {
+	pvName := req.GetPvName()
+	if pvName == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "pv name is missing from the request")
+	}
+
+	pv, err := ekrs.kubeClient.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to find a pv with name: %s. error: %v", pvName, err)
+	}
+
+	if pv.Spec.CSI == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "%s is not a CSI volume", pvName)
+	}
+
+	for _, mode := range pv.Spec.AccessModes {
+		if mode == corev1.ReadWriteMany || mode == corev1.ReadOnlyMany {
+			return nil, status.Errorf(codes.InvalidArgument, "%s is RWX/ROX volume, key rotation unsupported", pvName)
+		}
+	}
+
+	volID := pv.Spec.CSI.VolumeHandle
+	volAttrs := pv.Spec.CSI.VolumeAttributes
+
+	ekrRequest := &csiEncKeyRotation.EncryptionKeyRotateRequest{
+		VolumeId:   volID,
+		Parameters: volAttrs,
+	}
+
+	if pv.Spec.CSI.NodeStageSecretRef != nil {
+		ekrRequest.Secrets, err = kube.GetSecret(ctx, ekrs.kubeClient, pv.Spec.CSI.NodeStageSecretRef.Name, pv.Spec.CSI.NodeStageSecretRef.Namespace)
```

Comment:
can you check this one, if node stage secret is missing we need to use publish secret

Reply:
Both are optional values, since `NodeStageVolume` comes before `NodePublishVolume` in the order of things, it would be present in the cases where it is required. 

Can you tell me of such case where stage secret might be empty but publish secret wouldn't?